core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -113,122 +111,117 @@ class MetadataCache(brokerId: Int) extends Logging {
-    inReadLock(partitionMetadataLock) {
-      val nodes = aliveNodes.map { case (id, nodes) => (id, nodes.get(listenerName).orNull) }
-      def node(id: Integer): Node = nodes.get(id).orNull
-      val partitions = getAllPartitions()
-        .filter { case (_, state) => state.basePartitionState.leader != LeaderAndIsr.LeaderDuringDelete }
-        .map { case (tp, state) =>
-          new PartitionInfo(tp.topic, tp.partition, node(state.basePartitionState.leader),
-            state.basePartitionState.replicas.asScala.map(node).toArray,
-            state.basePartitionState.isr.asScala.map(node).toArray,
-            state.offlineReplicas.asScala.map(node).toArray)
-        }
-      val unauthorizedTopics = Collections.emptySet[String]
-      val internalTopics = getAllTopics().filter(Topic.isInternal).asJava
-      new Cluster(clusterId, nodes.values.filter(_ != null).toList.asJava,
-        partitions.toList.asJava,
-        unauthorizedTopics, internalTopics,
-        getControllerId.map(id => node(id)).orNull)
-    }
+    val snapshot = metadataSnapshot
+    val nodes = snapshot.aliveNodes.map { case (id, nodes) => (id, nodes.get(listenerName).orNull) }
+    def node(id: Integer): Node = nodes.get(id).orNull
+    val partitions = getAllPartitions(snapshot)
+      .filter { case (_, state) => state.basePartitionState.leader != LeaderAndIsr.LeaderDuringDelete }
+      .map { case (tp, state) =>
+        new PartitionInfo(tp.topic, tp.partition, node(state.basePartitionState.leader),
+          state.basePartitionState.replicas.asScala.map(node).toArray,
+          state.basePartitionState.isr.asScala.map(node).toArray,
+          state.offlineReplicas.asScala.map(node).toArray)
+      }
+    val unauthorizedTopics = Collections.emptySet[String]
+    val internalTopics = getAllTopics(snapshot).filter(Topic.isInternal).asJava
+    new Cluster(clusterId, nodes.values.filter(_ != null).toList.asJava,
+      partitions.toList.asJava,
+      unauthorizedTopics, internalTopics,
+      snapshot.controllerId.map(id => node(id)).orNull)
   }
 
   // This method returns the deleted TopicPartitions received from UpdateMetadataRequest
   def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest): Seq[TopicPartition] = {
     inWriteLock(partitionMetadataLock) {
-      controllerId = updateMetadataRequest.controllerId match {
+
+      //since kafka may do partial metadata updates, we start by copying the previous state
+      val cache = mutable.Map[String, mutable.Map[Int, UpdateMetadataRequest.PartitionState]]()
+      metadataSnapshot.partitionStates.foreach { case (topic, md) =>
+        val copy = mutable.Map[Int, UpdateMetadataRequest.PartitionState]()
+        copy ++= md
+        cache += (topic -> copy)
+      }
+      val aliveBrokers = mutable.Map[Int, Broker]()
+      val aliveNodes = mutable.Map[Int, collection.Map[ListenerName, Node]]()
```

Comment:
Can we allocate these maps with the correct size immediately? Also, we could save memory by using `LongMap` which uses open addressing.